HHH-18377 Support for uuid v6 and v7 generated ids

cigaly opened the pull request:
Jira issue [HHH-18377](https://hibernate.atlassian.net/browse/HHH-18377)

Implementation of [UUID Versioin 6](https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-6) and [UUID Version 7](https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-7)

<!--
Please read and do not remove the following lines:
-->
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------


[HHH-18377]: https://hibernate.atlassian.net/browse/HHH-18377?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

CaptainGlac1er commented:
How will it handle if a separate server generates the same id into the same database? Will it recreate a new ID automatically and try again?

sebersole commented:
> How will it handle if a separate server generates the same id into the same database? Will it recreate a new ID automatically and try again?

No.

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidV6ValueGenerator.java:
```diff
@@ -0,0 +1,126 @@
+import org.hibernate.id.UUIDGenerationStrategy;
+
+/**
+ * Implements a <a href="https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-6">UUID Version 6</a> generation strategy as defined by the {@link UUID#randomUUID()} method.
```
Not sure why you are referencing `UUID#randomUUID` here?

cigaly replied:
Can not explain that otherwise than some stupid mistake. Used wrong link and repeated that four times.

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidV6ValueGenerator.java:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	/**
+	 * Delegates to {@link UUID#randomUUID()}
```
Not sure why you are referencing `UUID#randomUUID` here?

cigaly replied:
Same mistake as on other three places

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidV7ValueGenerator.java:
```diff
@@ -0,0 +1,123 @@
+import static java.time.temporal.ChronoUnit.MILLIS;
+
+/**
+ * Implements a <a href="https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-7">UUID Version 7</a> generation strategy as defined by the {@link UUID#randomUUID()} method.
```
Not sure why you are referencing `UUID#randomUUID` here?

cigaly replied:
Same mistake as on other three places

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidV7ValueGenerator.java:
```diff
@@ -0,0 +1,123 @@
+	}
+
+	/**
+	 * Delegates to {@link UUID#randomUUID()}
```
Not sure why you are referencing `UUID#randomUUID` here?

cigaly replied:
Same mistake as on other three places

sebersole reviewed (changes requested):
I left a few comments about odd references to `UUID#randomUUID()`.  Let's clean those up.

Additionally, you write great detail about what each value generator does.  But you do it on `#generateUuid`.  I'd prefer to see those on the class Javadoc, because...

Another thing we should do here is to add values for these into `UuidGenerator.Style`.  Its probably best to have these link to the class Javadoc for the corresponding value-generator.

sebersole commented:
A few naming related points here...

1. Not a fan of the enum names (UUIDV6, UUIDV7).  How about VERSION6 and VERSION7 (or VERSION_6...)?  Or since they are RFC compliant, RFC_6 and RFC_7?
2. The value generator names should use `{Type}Strategy` pattern to match with the others.  E.g. `Version6Strategy`.

I'll need to pull this locally to squash the commits before I apply.  I can apply the name changes at that time if you prefer.

Thanks again for the work on this!

cigaly commented:
> A few naming related points here...
> 
>     1. Not a fan of the enum names (UUIDV6, UUIDV7).  How about VERSION6 and VERSION7 (or VERSION_6...)?  Or since they are RFC compliant, RFC_6 and RFC_7?
> 
>     2. The value generator names should use `{Type}Strategy` pattern to match with the others.  E.g. `Version6Strategy`.

Changed both things.

> I'll need to pull this locally to squash the commits before I apply. I can apply the name changes at that time if you prefer.

I've squashed everything into two commits, one for implementation, other for test case. Hope that this is OK?

> Thanks again for the work on this!

You're welcome. Happy if I can help. Not that I was too busy lately.

sebersole commented:
In general, we like the test/impl split for the PR as it shows (1) what is addressed and (2) how it was addressed.  

When we apply them though we generally (not always) squash to one as it makes it easier to cherry-pick to other branches.  This we won't be backporting so its less important.  So long story short, 2 commits is fine here.

CaptainGlac1er commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion7Strategy.java:
```diff
@@ -0,0 +1,123 @@
+		final long nanosPart = Math.round( ( currentTimestamp.getNano() % 1_000_000L ) * 0.004096 );
+
+		return new UUID(
+				millis << 16 & 0xFFFF_FFFF_FFFF_0000L | 0x7000L | nanosPart & 0xFFFL,
```
Feel fine to ignore this comment, but would it make it more self documenting to save this mask to a constant?

CaptainGlac1er replied:
I would advise against code comments, as code comments tend to trend to lie as the project ages. 
https://dev.to/ruben_alapont/magic-numbers-and-magic-strings-its-time-to-talk-about-it-ci2

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidGenerator.java:
```diff
@@ -81,6 +83,12 @@ private static UuidValueGenerator determineValueGenerator(
 			else if ( config.style() == TIME ) {
 				return new CustomVersionOneStrategy();
 			}
+			else if ( config.style() == VERSION_6 ) {
```
Looking at this closer now though, I wonder if we should reuse the same instance of these (singleton) rather than creating one for each id?

cigaly replied:
Perhaps, yes. This will prevent two entities to have equal id, but this is something that is good to avoid, anyway.

sebersole replied:
Looks great, thanks!

cigaly commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidGenerator.java:
```diff
@@ -81,6 +83,12 @@ private static UuidValueGenerator determineValueGenerator(
+			else if ( config.style() == VERSION_6 ) {
+				return new UuidVersion6Strategy();
+			}
+			else if ( config.style() == VERSION_7 ) {
```
Not sure if this will look any better than now ...

```
if ( config.algorithm() != UuidValueGenerator.class ) {
...
}
else {
    switch (config.style()) {
    ...
    }
}
```

CaptainGlac1er replied:
You won't need the `else` since the first if statement returns.

CaptainGlac1er replied:
Or like

```java
if(config == null) {
  return StandardRandomStrategy.INSTANCE;
}
if(config.algorithm() != UuidValueGenerator.class) {
  // throw or return logic
}

switch(config.style()) {
  case TIME:
    return ..
  case VERSION_6:
    return UuidVersion6Strategy.INSTANCE;
  case VERSION_7:
    return UuidVersion7Strategy.INSTANCE;
}
```

sebersole replied:
Well no.  You are missing what is effectively the AUTO, which also happens to  be the `config == null` case.

I'm not following the desire to add a switch just to add a switch.  A switch with so few case options is unuseful imo.  In fact, IntelliJ will even suggest this as a refactoring.

CaptainGlac1er replied:
Well was trying to flatten the if statement and moved it up as a guard clause for the AUTO. 
https://blog.codinghorror.com/flattening-arrow-code/

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion7Strategy.java:
```diff
@@ -0,0 +1,123 @@
+ *     <li>14 bits - counter to guarantee additional monotonicity, resets to 0 when timestamp changes. </li>
+ *     <li>48 bits - pseudorandom data to provide uniqueness.</li>
+ * </ul>
+ *
```
```suggestion
 *
 * @apiNote A Version 7 UUID features a time-ordered value field derived from the widely
 * implemented and well known Unix Epoch timestamp source, the number of milliseconds
 * since midnight 1 Jan 1970 UTC, leap seconds excluded.
 *
```

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion7Strategy.java:
```diff
@@ -0,0 +1,123 @@
+	 */
+	@Override
+	public int getGeneratedVersion() {
+		/*
+		 *	UUIDv7 features a time-ordered value field derived from the widely implemented and well-
+		 * known Unix Epoch timestamp source, the number of milliseconds since midnight 1 Jan 1970 UTC,
+		 * leap seconds excluded.
+		 */
```
```suggestion
```

sebersole commented on hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion7Strategy.java:
```diff
@@ -0,0 +1,123 @@
+	}
+
+	/**
+	 * @param session session
+	 *
+	 * @return UUID version 7
+	 * @see UuidValueGenerator#generateUuid(SharedSessionContractImplementor)
```
```suggestion
	 * Generates a Version 7 {@linkplain UUID}.
	 *
	 * @return A Version version 7 {@linkplain UUID}.
	 * @see UuidValueGenerator#generateUuid(SharedSessionContractImplementor)
```

sebersole reviewed (commented):
A few minor suggestions/requests.  Let me know what you think.  FWIW I left "suggested changes" in those places to make it easy to apply if you agree.

sebersole commented:
Also, clearly we need to address the failures :)

```console
Execution failed for task ':hibernate-core:compileTestJava'.
> Compilation failed; see the compiler error output for details.
```

cigaly commented:
> Also, clearly we need to address the failures :)
> 
> ```
> Execution failed for task ':hibernate-core:compileTestJava'.
> > Compilation failed; see the compiler error output for details.
> ```

One will expect that IDE should know to make basic refactoring things like renaming classes :-)

sebersole commented:
Applied manually.  Thanks!

beikov commented:
I'm getting test errors quite consistently. I didn't have a single success of the v6 test so far:

```
java.lang.AssertionError: 
Expecting actual:
  "1ef7be82-786f-69e5-8000-5b0eb9699e10"
to be greater than:
  "1ef7be82-786f-69e5-bfff-e4cda4de4e50"

	at org.hibernate.orm.test.id.uuid.rfc9562.UUidV6V7GenetartorTest.testMonotonicity(UUidV6V7GenetartorTest.java:43)
	at org.hibernate.orm.test.id.uuid.rfc9562.UUidV6V7GenetartorTest.testMonotonicityUuid6(UUidV6V7GenetartorTest.java:26)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

Running with jdk-17.0.12+7 on Windows 11 if it matters. I don't know how the implementation works, but it seems broken.

cigaly commented:
It seems that there is pretty stupid bug :-( I'll check and fix that.

beikov commented:
I just ran into another (unrelated but possibly similar) error, which is related to the fact that on Windows, the time source has a precision of 7 digits after the decimal separator.

cigaly commented:
I've located that problem and will fix it ASAP

cigaly commented:
@beikov I've created new pull request [#9020](https://github.com/hibernate/hibernate-orm/pull/9020) with updates to both UUID v6 and v7 generation
